pystorm/spout.py

```diff
@@ -97,3 +91,76 @@ def _run(self):
         else:
             self.logger.error('Received invalid command from Storm: %r', cmd)
         self.send_message({'command': 'sync'})
+
+
+class ReliableSpout(Component):
+    """Reliable spout that will automatically replay failed tuples.
+
+    Failed tuples will be replayed up to ``max_fails`` times.
+
+    For more information on spouts, consult Storm's
+    `Concepts documentation <http://storm.apache.org/documentation/Concepts.html>`_.
+    """
+    max_fails = 3
+
+    def __init__(self, *args, **kwargs):
+        super(ReliableSpout, self).__init__(*args, **kwargs)
+        self.failed_tuples = Counter()
+        self.unacked_tuples = {}
+
+    def ack(self, tup_id):
+        """Called when a bolt acknowledges a Tuple in the topology.
+
+        :param tup_id: the ID of the Tuple that has been fully acknowledged in
+                       the topology.
+        :type tup_id: str
+        """
+        if tup_id in self.failed_tuples:
+            del self.failed_tuples[tup_id]
+        if tup_id in self.unacked_tuples:
+            del self.unacked_tuples[tup_id]
+        else:
+            self.logger.error('Received ack for unknown tuple ID: %r', tup_id)
+
+    def fail(self, tup_id):
+        """Called when a Tuple fails in the topology
+
+        A reliable spout will replay a failed tuple up to ``max_fails`` times.
+
+        :param tup_id: the ID of the Tuple that has failed in the topology
+                       either due to a bolt calling ``fail()`` or a Tuple
+                       timing out.
+        :type tup_id: str
+        """
+        saved_args = self.unacked_tuples.get(tup_id)
+        if saved_args is None:
+            self.logger.error('Received fail for unknown tuple ID: %r', tup_id)
+            return
+        tup, stream, direct_task, need_task_ids = saved_args
+        if self.failed_tuples[tup_id] < self.max_fails:
+            self.emit(tup, tup_id=tup_id, stream=stream,
+                      direct_task=direct_task, need_task_ids=need_task_ids)
+            self.failed_tuples[tup_id] += 1
+        else:
+            # Just pretend we got an ack when we exceed retry limit
+            self.logger.info('Acking tuple ID %r after it exceeded retry limit '
+                             '(%r)', tup_id, self.max_fails)
+            self.ack(tup_id)
```

Comment:
Spouts do have an `ack` method. It gets called when a tuples gets acked so that we can update the bookkeeping. It's actually the method your `pop` comment was on. 😄 

By just calling `ack` here, we make it so people can override `ack` if they want to do anything more complex, and it'll clear the things from `failed_tuples` and `unacked_tuples` just like when we get a real ack message.